File: 3.0/python/38.py
N, M, S, T, Q = map(int, input().split())
S, T = S-1, T-1

field = [[-1]*M for _ in range(N)]
field[S][T] = 0
res = {0: [(T, S)]}

dx = [-2, -2, -1, -1, 1, 1, 2, 2]
dy = [1, -1, 2, -2, 2, -2, 1, -1]

for i in range(1, N * M):
    res[i] = []
    for coords in res[i - 1]:
        for delta in range(8):
            x = coords[0] + dx[delta]
            y = coords[1] + dy[delta]
            if 0 <= x < M and 0 <= y < N:
                if field[y][x] == -1:
                    field[y][x] = i
                    new = (x, y)
                    res[i].append(new)

ans = 0
for _ in range(Q):
    y, x = map(int, input().split())
    x -= 1
    y -= 1
    if field[y][x] == -1:
        ans = -1
        break
    ans += field[y][x]
print(ans)

File: 3.0/golang/38.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var N, M, S, T, Q int
	in := bufio.NewReader(os.Stdin)
	fmt.Fscanf(in, "%d %d %d %d %d\n", &N, &M, &S, &T, &Q)
	S, T = S-1, T-1

	field := make([][]int, N)
	for i := range field {
		for j := 0; j < M; j++ {
			field[i] = append(field[i], -1)
		}
	}
	field[S][T] = 0
	res := make(map[int][][]int)
	res[0] = [][]int{{T, S}}

	dx := []int{-2, -2, -1, -1, 1, 1, 2, 2}
	dy := []int{1, -1, 2, -2, 2, -2, 1, -1}

	for i := 1; i < N*M; i++ {
		res[i] = [][]int{}
		for _, coords := range res[i-1] {
			for delta := 0; delta < 8; delta++ {
				x := coords[0] + dx[delta]
				y := coords[1] + dy[delta]
				if 0 <= x && x < M && 0 <= y && y < N {
					if field[y][x] == -1 {
						field[y][x] = i
						res[i] = append(res[i], []int{x, y})
					}
				}
			}
		}
	}

	ans := 0
	var y, x int
	for i := 0; i < Q; i++ {
		fmt.Fscanf(in, "%d %d\n", &y, &x)
		x--
		y--
		if field[y][x] == -1 {
			ans = -1
			break
		}
		ans += field[y][x]
	}
	fmt.Println(ans)
}
